### PR TITLE
Improve in_footer handling in gutenberg_override_script()

### PR DESCRIPTION
**Pull request by ocean90**
## What?
This is related to https://meta.trac.wordpress.org/ticket/6195 ~but still doesn't fix the issue.~ and adds a temporary workaround for a broken script dependency issue related to `common` and `wp-i18n`.

## Why?
`$args` should be set to `1` if a script should be loaded in the footer, see https://github.com/WordPress/wordpress-develop/blob/573581f43f63f12cc37620172432cf93745f568c/src/wp-includes/class.wp-scripts.php#L541-L542.

### CODE REVIEW

File: lib/client-assets.php [hunk 1]
```diff
@@ -66,22 +66,9 @@ function gutenberg_override_script( $scripts, $handle, $src, $deps = array(), $v
...
-
-		// See: `wp_register_script` .
-		unset( $script->extra['group'] );
-		if ( $in_footer ) {
-			$script->add_data( 'group', 1 );
-		}
+		$script->args = $in_footer ? 1 : null;
 	} else {
-		$scripts->add( $handle, $src, $deps, $ver, $in_footer );
+		$scripts->add( $handle, $src, $deps, $ver, ( $in_footer ? 1 : null ) );
```
Comment (aristath) [inline lib/client-assets.php:77]: A bit of a nitpick, but can we use `true`/`false` instead of `1`/`null`? I know that they technically do the same thing, and we wouldn't have any issues using 1/null here... but newer PHP versions are a bit more strict when it comes to types, so just to be safe for the future we should probably stick to boolean values. 👍

Comment (ocean90) [reply lib/client-assets.php:77]: No, `null` is the default value and only `1` is supported by the [linked special case](https://github.com/WordPress/wordpress-develop/blob/573581f43f63f12cc37620172432cf93745f568c/src/wp-includes/class.wp-scripts.php#L541-L542).

**Review by youknowriad — APPROVED** [review]

File: lib/client-assets.php (hunk 1 above)
Comment (aristath) [reply lib/client-assets.php:77]: Ah you're right... my bad! Thanks for pointing me in the right direction 👍

**Review by aristath — APPROVED** [review]
LGTM! 👍

File: lib/client-assets.php [hunk 2]
```diff
@@ -66,22 +66,9 @@ function gutenberg_override_script( $scripts, $handle, $src, $deps = array(), $v
...
 	}
 
 	/*
```
Comment (youknowriad) [inline lib/client-assets.php:80]: It doesn't solve the issue either but I noticed that some gutenberg scripts receive the extra wp-i18n even if they don't need it (like wp-deprecated). We should do like core and only call `set_translations` if `wp-i18n` is part of the $dependencies array.

**Comment by youknowriad** [comment]
I confirmed that after the `gutenberg_register_packages_scripts` function is executed, the `common` script was untouched. But I also saw that it doesn't have any `wp-i18n` dependency even if I see in core a call to `set_translations` which is supposed to add this dependence (and fix the loading order).

Gutenberg though potentially changes loading order of some scripts and that's normal since the dependencies evolve in Gutenberg packages compared to Core. I suspect that the current bug is a side effect of a legitimate change here.

In other words, I think the main issue is that "common" doesn't have a dependency to "wp-i18n" in Core, and that by luck, it was loaded before in Core in 5.9 and before but for in Gutenberg now it gets loaded after. We should try to find why "common" script doesn't have "wp-i18n" in its deps even if `set_translations` is called on the script loader.

**Comment by ocean90** [comment]
The E2E tests did pass in #39501 so I think this can be merged now.

**Comment by hellofromtonya** [comment]
Wiring slack conversations here for context.

Per @ocean90 ([see slack](https://wordpress.slack.com/archives/C02QB2JS7/p1647440786601139?thread_ts=1647440087.132519&cid=C02QB2JS7)):
>39497 fixes the registration of a script for head/footer and 39501 implements a workaround to register wp-i18n by default in head.
>
>39497 now includes 39501

**Comment by jsnajdr** [comment]
> We should try to find why "common" script doesn't have "wp-i18n" in its deps even if set_translations is called on the script loader.

@youknowriad That's because when you call `$scripts->set_translations( 'common' )`, the `set_translations` method will automatically add the `wp-i18n` dependency:

https://github.com/jsnajdr/wordpress-develop/blob/trunk/src/wp-includes/class.wp-scripts.php#L573-L575

I found out today that the issue is somewhere else: in the way how `WP_Scripts` concatenates scripts with `load-scripts.php`. Scripts that are in one of the default directories (`wp-includes/js` etc) are handled differently than ones in a non-default directory. And when a script in a default directory depends on a script in a non-default directory, things break down and scripts are printed out of order.

When Gutenberg registers its own version of `wp-i18n`, exactly this situation arises: `common` (default dir) depends on `wp-i18n` (non-default dir).

Now I also understand why removing empty script translations in https://core.trac.wordpress.org/ticket/55250 suddenly triggered these errors. "Script has translations" is one of the conditions that disqualifies a script from concatenation. After this patch, suddenly many more scripts are getting concatenated.

I documented my investigations also in https://core.trac.wordpress.org/ticket/55628#comment:15

Cc also @ocean90
